src/som/interpreter/actors/Actor.java

```diff
@@ -182,11 +187,34 @@ protected void appendToMailbox(final EventualMessage msg) {
     mailboxExtension.append(msg);
   }
 
-  protected static void handleBreakPoints(final EventualMessage msg, final WebDebugger dbg) {
-    if (VmSettings.TRUFFLE_DEBUGGER_ENABLED && msg.isBreakpoint()) {
-      dbg.prepareSteppingUntilNextRootNode();
+  protected static void handleBreakpointsAndStepping(final EventualMessage msg, final WebDebugger dbg) {
+    if (VmSettings.TRUFFLE_DEBUGGER_ENABLED) {
+      // check step-into and breakpoint before the message is processed
+      if (msg.isStepInto()) {
+        dbg.prepareStepIntoMessage();
+      } else if (msg.isBreakpoint()) {
+        dbg.prepareSteppingUntilNextRootNode();
+      }
+      // check if a step-over has been made in the previous message
+      if (stopNextMessage) {
+       dbg.prepareStepIntoMessage();
+      }
+      // check step-return
+      if (msg.isStepReturn()) {
+          dbg.prepareStepReturnMessage(msg.getCausalMessage());
+      }
+      // check type of callback if a step-return has been made
+      if (msg instanceof PromiseSendMessage) {
+        if (((PromiseSendMessage) msg).getPromise().isTriggerStepReturnOnCallbacks()) {
```

Comment:
instead of having a comment, I think it would be better to have a helper method that does the check, and where the name represents what it is supposed to do, i.e., encodes the essence of the comment.